src/helpers/__tests__/upload.test.ts

```diff
@@ -0,0 +1,193 @@
+      expect(uploadCallback).toBeCalledTimes(1)
+      expect(errorCallback).toBeCalledTimes(1)
+      expect(retryCallback).toBeCalledTimes(0)
+      expect(result).toStrictEqual(UploadStatus.Failure)
+    })
+
+    test('should retry only a given amount of times', async () => {
+      const mockCreate = mockAxiosResponse([
+        () =>
+          Promise.reject({
+            response: {
+              status: 413,
+            },
+          }),
+        () =>
+          Promise.reject({
+            response: {
+              status: 413,
+            },
+          }),
+      ])
+
+      const request = ciUtils.getRequestBuilder({apiKey: '', baseUrl: ''})
+
+      const result = await upload(request)(
+        {content: new Map()},
+        {
+          onError: errorCallback,
+          onRetry: retryCallback,
+          onUpload: uploadCallback,
+          retries: 1,
+        }
+      )
+      expect(mockCreate).toBeCalledTimes(1)
+      expect(uploadCallback).toBeCalledTimes(1)
+      expect(errorCallback).toBeCalledTimes(1)
+      expect(retryCallback).toBeCalledTimes(0)
+      expect(result).toStrictEqual(UploadStatus.Failure)
+    })
+
+    test('apiKeyValidator should not be called in case of success', async () => {
+      const mockCreate = mockAxiosResponse([() => Promise.resolve({})])
+      const request = ciUtils.getRequestBuilder({apiKey: '', baseUrl: ''})
+      verifyKey.mockImplementation(() => Promise.resolve())
+      const result = await upload(request)(
+        {content: new Map()},
+        {
+          apiKeyValidator: {
+            verifyApiKey: verifyKey,
+          },
+          onError: errorCallback,
+          onRetry: retryCallback,
+          onUpload: uploadCallback,
+          retries: 1,
+        }
+      )
+      expect(mockCreate).toBeCalledTimes(1)
+      expect(verifyKey).toBeCalledTimes(0)
+      expect(uploadCallback).toBeCalledTimes(1)
+      expect(errorCallback).toBeCalledTimes(0)
+      expect(retryCallback).toBeCalledTimes(0)
+      expect(result).toStrictEqual(UploadStatus.Success)
+    })
+
+    test('apiKeyValidator should be called in case of ambiguous response', async () => {
+      const mockCreate = mockAxiosResponse([
+        () =>
+          Promise.reject({
+            response: {
+              status: 400,
+            },
+          }),
+      ])
+      const request = ciUtils.getRequestBuilder({apiKey: '', baseUrl: ''})
+      verifyKey.mockImplementation(() => Promise.reject('errorApiKey'))
+      const result = upload(request)(
+        {content: new Map()},
+        {
+          apiKeyValidator: {
+            verifyApiKey: verifyKey,
+          },
+          onError: errorCallback,
+          onRetry: retryCallback,
+          onUpload: uploadCallback,
+          retries: 1,
+        }
+      )
+      expect(result).rejects.toMatch('errorApiKey')
+      expect(mockCreate).toBeCalledTimes(1)
+      expect(uploadCallback).toBeCalledTimes(1)
```

Comment:
I was about to suggest to add `expect(verifyKey).toBeCalledTimes(1)` but it doesn't seem to work, no idea why.

Reply:
Yes I initially added this check but couldn't make it work, I wonder if this is related to how exception/rejected promise handling work with jest. I didn't dig further because I realized if the rejected promise matched `errorApiKey` that proves that the mocked `verifyKey` function has been called.